src/include/duckdb/common/lru_cache.hpp

```diff
@@ -0,0 +1,160 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/lru_cache.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <functional>
+#include <list>
+
+#include "duckdb/common/common.hpp"
+#include "duckdb/common/hash_utils.hpp"
+#include "duckdb/common/mutex.hpp"
+#include "duckdb/common/shared_ptr.hpp"
+#include "duckdb/common/string.hpp"
+#include "duckdb/common/unordered_map.hpp"
+
+namespace duckdb {
+
+// A LRU cache implementation, whose value could be accessed in a shared manner with shared pointer.
+// Notice, it's not thread-safe.
+template <typename Key, typename Val, typename KeyHash = std::hash<Key>, typename KeyEqual = std::equal_to<Key>>
+class SharedLruCache {
```

Comment:
Instead of templating with 4 parameters, I think it makes more sense to template with the entire map type. Then there is only one template parameter, and this class becomes potentially more flexible (e.g., you can use it with an `ordered_map` if you want, or a `case_insensitive_map_t`).

Then you can still do this for convenience:
```c++
template <typename MAP_TYPE>
class SharedLRUCache {
public:
    using key_type = MAP_TYPE::key_type;
    // etc.
};
```
And then you can template this with `reference_map<string, shared_ptr<ObjectCacheEntry>>` in `object_cache.hpp`.